cosmetics-web/spec/controllers/responsible_persons/notifications/nanomaterials/build_controller_spec.rb

```diff
@@ -73,32 +81,58 @@
+          expect(nano_material1.reload.purposes).to eq(%w[preservative uv_filter])
+        end
+
+        it "ignores invalid purpose values" do
+          post(:update, params: select_purposes_params.merge(purposes_form: { "colorant": "1", "invalid_purpose": "1", "purpose_type": "standard" }))
+          expect(nano_material1.reload.purposes).to eq(%w[colorant])
+        end
+
+        it "redirects to the next page when purposes are selected" do
+          post(:update, params: select_purposes_params.merge(purposes_form: { "preservative": "1", "purpose_type": "standard" }))
+          expect(response).to redirect_to(
+            responsible_person_notification_nanomaterial_build_path(responsible_person, notification, nano_material1, :after_select_purposes_routing),
+          )
+        end
+
+        it "sets error when no purpose type is selected" do
+          post(:update, params: select_purposes_params)
+          expect(assigns(:purposes_form).errors[:purpose_type]).to include("Select the purpose of this nanomaterial")
+        end
+
+        it "sets error when no purpose is selected for standard purpose type" do
+          post(:update, params: select_purposes_params.merge(purposes_form: { "purpose_type": "standard", "colorant": "0", "preservative": "0", "uv_filter": "0" }))
+          expect(assigns(:purposes_form).errors[:purposes]).to include("Select the purpose")
+        end
+
+        it "correctly switches from a standard to a non-standard nanomaterial" do
+          post(:update, params: select_purposes_params.merge(purposes_form: { "purpose_type": "other" }))
+          expect(nano_material1.reload.purposes).to eq(%w[other])
+          expect(response).to redirect_to(
+            responsible_person_notification_nanomaterial_build_path(responsible_person, notification, nano_material1, :after_select_purposes_routing),
+          )
+        end
       end
 
-      it "sets error when no purpose is selected for standard purpose type" do
-        post(:update, params: select_purposes_params.merge(purposes_form: { "purpose_type": "standard", "colorant": "0", "preservative": "0", "uv_filter": "0" }))
-        expect(assigns(:purposes_form).errors[:purposes]).to include("Select the purpose")
+      context "with a non-standard nanomaterial" do
+        it "sets error when no purpose type is selected" do
+          post(:update, params: select_purposes_params_non_standard)
+          expect(assigns(:purposes_form).errors[:purpose_type]).to include("Select the purpose of this nanomaterial")
+        end
+
+        it "correctly switches from a non-standard to a standard nanomaterial" do
+          post(:update, params: select_purposes_params_non_standard.merge(purposes_form: { "colorant": "0", "preservative": "1", "uv_filter": "1", "purpose_type": "standard" }))
+          expect(nano_material2.reload.purposes).to eq(%w[preservative uv_filter])
```

Comment:
I guess we want to check that the association with the `nanomaterial_notification` is no longer present.
```suggestion
          expect(nano_material2.reload).to have_attributes(purposes: %w[preservative uv_filter], nanomaterial_notification_id: nil)
```